### ISSUE TITLE
`node::net.Socket()` reusing socket works on NodeJS but fails in Bun

### ISSUE THREAD

**Issue by wackfx** [oven-sh/bun#7325]
### What version of Bun is running?

1.0.13

### What platform is your computer?

Darwin 22.6.0 arm64 arm

### What steps can reproduce the bug?

Reusing a client socket connection works in Node.JS - but fails in Bun.

Reusing means that we close the connection at some point for some reasons then we try to reconnect and reuse the socket.
This use case is often implemented in database clients to avoid stale socket issues.

Test script - you should have a simple echo TCP server running
```javascript
// Run tcp server with `docker run --rm -d -p 9000:9000 venilnoronha/tcp-echo-server:latest`
import net from 'node:net'

// Pretty standard socket stuff & wait util
const socket = new net.Socket()
function on_data(data) {
    console.log('data', data.toString())
}
function bind_listeners() {
    socket.on('data', on_data)
}
function remove_listeners() {
    socket.removeListener('data', on_data)
}
async function wait(ms) {
    return new Promise((ok, ko) => {
        setTimeout(ok, ms)
    })
}

// Actual test code
// we want to simulate a standard 1) socket connection 2) writing data 3) eventually close
// In a real world scenario, the socket would end much later (in a setTimeout for ex)
// but for this exemple to have the error faster, we trigger the end right after we wrote some data
// NB: It still give an error even if you do a longer period before connect(ing) and end(ing)
async function send_hello() {
    return new Promise((ok, ko) => {
        // binding event - connecting's done at the end
        socket.once('connect', (...args) => {
            console.log('connected')
            // Binding data listener - NodeJS fails without it (might be a lead for some shenanigans?)
            bind_listeners()
            socket.write('script\n', (err) => {
                if (err) return ko(err)
                console.log('wrote')
                // remove added listeners & end connection
                remove_listeners()
                socket.end(ok)
            })
        })
        socket.connect(9000, '127.0.0.1')
    })
}

// Try to perform this action 3 times
async function main() {
    await send_hello()
    await wait(1000)
    await send_hello()
    await wait(1000)
    await send_hello()
}
await main()

```

### What is the expected behavior?

No error (like in node)
![image](https://github.com/oven-sh/bun/assets/135170502/a47ec7fa-73cb-4271-9f1c-bfd5bac869e9)
![image](https://github.com/oven-sh/bun/assets/135170502/f5ac8fd6-68ee-4497-9689-03b88b8828c0)

### What do you see instead?

It fails with Bun
![image](https://github.com/oven-sh/bun/assets/135170502/26b1ef49-99d3-4c63-ab05-a7f239f211bc)

### Additional information

Found this while debugging a [pretty famous node module](https://github.com/porsager/postgres/pull/738) that froze over a period of time in Bun but worked flawlessly on Node.JS.

A solution we found to bypass this error was to recreate a `new.Socket()` each time after a socket "ended" (instead of reusing the same socket)

**Comment by deanrih** [comment]
It seem this is still the issue on the latest version `1.0.27`

```
❯ node --version; node ../test.js
v20.10.0
connected
wrote
connected
wrote
connected
wrote
```

```
❯ bun --version; bun run ../test.js
1.0.27
connected
wrote
connected
28 |         // binding event - connecting's done at the end
29 |         socket.once('connect', (...args) => {
30 |             console.log('connected')
31 |             // Binding data listener - NodeJS fails without it (might be a lead for some shenanigans?)
32 |             bind_listeners()
33 |             socket.write('script\n', (err) => {
                 ^
error: write after end
 code: "ERR_STREAM_WRITE_AFTER_END"

      at new NodeError (node:stream:405:39)
      at _write (node:stream:2546:39)
      at node:stream:2579:79
      at /mnt/tier_2_common/workspace/test/test.js:33:13
      at open (node:net:74:17)
```

**Comment by evelant** [comment]
For anyone else who lands here with an issue using postgres-js on Bun, `3.4.4` will not release the connection when you call `end`. You can roll back to `3.4.3` which will release the connection but will fail after ~1hr or so if you keep it open.

**Comment by wiesys** [comment]
Hmm... the issue with Postgres.js not properly closing the connection for me is present even in Bun v1.1.8. Are you sure @Jarred-Sumner that it is fixed?

Looks like I'm not the only one: https://github.com/porsager/postgres/issues/817#issuecomment-2117616412

**Comment by amjed-ali-k** [comment]
Postgres.js 3.4.3 | Bun 1.1.7 | 20 Mins
![image](https://github.com/oven-sh/bun/assets/86785660/1f48d7ca-3401-48bc-847e-858932303f0d)

Postgres.js 3.4.4 | Bun 1.1.7 | 15 Mins
![image](https://github.com/oven-sh/bun/assets/86785660/36a13039-5ada-4e7b-b918-b3ab89927fb9)

**Comment by deanrih** [comment]
@amjed-ali-k 

Hi, may I have the code you are using to test it and also perhaps the postgresql config if any (maybe such us timeout and such), I want to test it on my end (which I know probably behave the same) and also trying some workaround I'm currently trying as well to validate it

**Comment by amjed-ali-k** [comment]
> @amjed-ali-k
> 
> Hi, may I have the code you are using to test it and also perhaps the postgresql config if any (maybe such us timeout and such), I want to test it on my end (which I know probably behave the same) and also trying some workaround I'm currently trying as well to validate it
> 
> Also can confirm, for me it has higher chance of the issue being triggered when the server was idling

@deanrih 
Created a fork from one of your repo. 
https://github.com/amjed-ali-k/bun-elysia-drizzle-postgres-test

Add your posgres db url in .env file.
run `bun run db:generate` & `bun run db:push`

Here is the screenshot of requests i made to the server. Check the time difference also. DB read request after 25 Mins got me this error.
![image](https://github.com/oven-sh/bun/assets/86785660/3d943d30-3e05-44d6-933a-6c91c1723b1e)

Also, i didn't got any error in locally hosted prostres. 
You can find my pg_config dump in [db-config-dump.txt](https://github.com/amjed-ali-k/bun-elysia-drizzle-postgres-test/blob/main/db-config-dump.txt)

**Comment by deanrih** [comment]
> > @amjed-ali-k
> > Hi, may I have the code you are using to test it and also perhaps the postgresql config if any (maybe such us timeout and such), I want to test it on my end (which I know probably behave the same) and also trying some workaround I'm currently trying as well to validate it
> > Also can confirm, for me it has higher chance of the issue being triggered when the server was idling
> 
> @deanrih Created a fork of one of your repo. https://github.com/amjed-ali-k/bun-elysia-drizzle-postgres-test
> 
> Add your posgres db url in .env file. run `bun run db:generate` & `bun run db:push`
> 
> Here is the screenshot of requests i made to the server. Check the time difference also. DB read request after 25 Mins got me this error. ![image](https://private-user-images.githubusercontent.com/86785660/333002483-3d943d30-3e05-44d6-933a-6c91c1723b1e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY0NjAxNTAsIm5iZiI6MTcxNjQ1OTg1MCwicGF0aCI6Ii84Njc4NTY2MC8zMzMwMDI0ODMtM2Q5NDNkMzAtM2UwNS00NGQ2LTkzM2EtNmM5MWMxNzIzYjFlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTIzVDEwMjQxMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTUxMjM5NTQ1MzBlZWE4NDY1YTkwOWFiMzA3NTRmNmMzY2YxNjg4NWE5NTExMzNjMjcxNGM0NGU1NTZlYzExZjEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.iCPjsQfqyxexoST9nGxkLf8-ErX0_7Pyki_06kuNhqY)
> 
> Also, i didn't got any error in locally hosted prostres. (for 30 mins) You can find my pg_config dump in [db-config-dump.txt](https://github.com/amjed-ali-k/bun-elysia-drizzle-postgres-test/blob/main/db-config-dump.txt)

@amjed-ali-k 

Oh wow, sorry for the late reply, and thank you, didn't expect for my dummy test repo to be used and kinda forgot about it (playing around with the local copy), will try it soon

**Comment by amjed-ali-k** [comment]
This issue seems fixed in bun 1.1.39 👍🏼
